### ISSUE TITLE
wxInfoBarGeneric hides the button if the width is small enough.

### ISSUE THREAD

**Issue by dsa-t** [wxWidgets/wxWidgets#24695]
#### To Reproduce:

1. Open `dialogs` sample
2. Click on Dialogs -> Informative Dialogs -> Simple info bar
3. Reduce the window width
4. The button is not visible anymore:
![image](https://github.com/wxWidgets/wxWidgets/assets/37658952/73857e10-959f-4664-8b19-b08671e3cadd)

---

1. Open `dialogs` sample
2. Click on Dialogs -> Informative Dialogs -> Simple info bar with wrapped text
3. The button is not visible:
![image](https://github.com/wxWidgets/wxWidgets/assets/37658952/4e82056f-0a3c-4bb1-8942-cbc39f8a7fa0)

---

Patch to make it work:

```patch
diff --git a/src/generic/infobar.cpp b/src/generic/infobar.cpp
index 1255d8961e..25f69a67d6 100644
--- a/src/generic/infobar.cpp
+++ b/src/generic/infobar.cpp
@@ -87,8 +87,7 @@ bool wxInfoBarGeneric::Create(wxWindow *parent, wxWindowID winid)
     //     and being preceded by a spacer
     wxSizer * const sizer = new wxBoxSizer(wxHORIZONTAL);
     sizer->Add(m_icon, wxSizerFlags().Centre().Border());
-    sizer->Add(m_text, wxSizerFlags().Centre());
-    sizer->AddStretchSpacer();
+    sizer->Add(m_text, wxSizerFlags().Centre().Proportion(1));
     sizer->Add(m_button, wxSizerFlags().Centre().Border());
     SetSizer(sizer);
 
@@ -233,7 +232,16 @@ void wxInfoBarGeneric::ShowMessage(const wxString& msg, int flags)
     // notice the use of EscapeMnemonics() to ensure that "&" come through
     // correctly
     m_text->SetLabel(wxControl::EscapeMnemonics(msg));
-    m_text->Wrap( GetClientSize().GetWidth() );
+
+    int width = GetClientSize().GetWidth();
+
+    if ( m_icon )
+        width -= m_icon->GetBestSize().y + wxSizerFlags::GetDefaultBorder() * 2;
+
+    if ( m_button )
+        width -= m_button->GetBestSize().y + wxSizerFlags::GetDefaultBorder() * 2;
+
+    m_text->Wrap(width);
 
     // then show this entire window if not done yet
     if ( !IsShown() )

```

![image](https://github.com/wxWidgets/wxWidgets/assets/37658952/85977968-8d6a-4ee5-8559-780bdc5f1c1c)

Note that `GetClientSize()` returns 20 for the first shown infobar, so the wrapping doesn't (and didn't) work at all on MSW.

Ideally it should re-wrap on size events (which requires storing the non-wrapped label AFAICS).
Would be nice to ellipsize the text also, if can't wrap.

---

KiCad issue: https://gitlab.com/kicad/code/kicad/-/issues/18359

### Platform and version information

- wxWidgets version  you use: 3.2.5 / master
- wxWidgets port  you use: wxMSW / any
- OS  and its version: Windows 11

**Comment by vadz** [comment]
Thanks for reporting this!

The first part of the patch looks good and I don't think we need anything else.

But wrapping was definitely implemented incorrectly in 32d8b5954a1. I guess it could be fixed by calling `Wrap()` after `Layout()` but I wonder if we could finally add (because this is not the first time it comes up) some `wxAutoWrappingLabel` which would call `Wrap()` with the current width whenever the label is resized.

**Comment by vadz** [comment]
Fixing wrapping is non-trivial because we need to change the height of the infobar in response to changes to its width. And this would have to be done in both generic (doable) and GTK (less obviously so) versions, as the latter is broken too currently.

So for now I've created #24838 which ellipsizes the message instead of wrapping it, as this is much simpler and can be easily backported to 3.2 too. Please let me know if you see any problems with it!
